no_redundant_asmjs_exports

juj opened the pull request:
Generate asm.js exports in form that does not reference the unminified exported name more than once, for smaller code size.

That is, instead of generating

```javascript
var _bitshift64Lshr = Module["_bitshift64Lshr"] = asm["_bitshift64Lshr"];
var _bitshift64Shl = Module["_bitshift64Shl"] = asm["_bitshift64Shl"];
var _free = Module["_free"] = asm["_free"];
var _i64Add = Module["_i64Add"] = asm["_i64Add"];
var _i64Subtract = Module["_i64Subtract"] = asm["_i64Subtract"];
var _llvm_bswap_i32 = Module["_llvm_bswap_i32"] = asm["_llvm_bswap_i32"];
var _main = Module["_main"] = asm["_main"];
var _malloc = Module["_malloc"] = asm["_malloc"];
var _memcpy = Module["_memcpy"] = asm["_memcpy"];
var _memmove = Module["_memmove"] = asm["_memmove"];
var _memset = Module["_memset"] = asm["_memset"];
var _sbrk = Module["_sbrk"] = asm["_sbrk"];
var _thisIsAFunctionExportedFromAsmJsOrWasmWithVeryLongFunctionNameThatWouldBeGreatToOnlyHaveThisLongNameReferredAtMostOnceInOutput = Module["_thisIsAFunctionExportedFromAsmJsOrWasmWithVeryLongFunctionNameThatWouldBeGreatToOnlyHaveThisLongNameReferredAtMostOnceInOutput"] = asm["_thisIsAFunctionExportedFromAsmJsOrWasmWithVeryLongFunctionNameThatWouldBeGreatToOnlyHaveThisLongNameReferredAtMostOnceInOutput"];
```
it should be equivalent to generate
```javascript
(function() { for(var i in asm) this[i] = Module[i] = asm[i] }());
```

after which we will only see the unminified name of an exported function appear only once, at the time it is returned as exports from the asm.js module.

For Wasm we have three references to the generated unminified function name versus asm.js one, but leaving that for later to tackle.

(I could have sworn we had this kind of transformation already land years ago, or recall seeing this kind of conversation occur, but perhaps the optimization never landed?)

kripken commented:
I think we did discuss this, but it didn't land due to either or both of

 * Test failures (I do see some now, like `test_ccall`)
 * Startup time regression. The loop might be slower than the straightline code. We should measure to be sure.

Other than those issues, I'm in favor of this.

kripken commented on src/closure-externs.js:
```diff
@@ -1370,3 +1370,40 @@ var onmessage = function(message) {};
+/**
+ * @suppress {duplicate, undefinedVars}
+ */
+var _malloc;
```
there should be nothing special about malloc and free that requires them to be here - this seems like it's trying to work around a remaining issue, that will occur with other functions?

kripken replied:
I think that's just by chance, though - there could be other functions not covered by the test suite that would be hit by this. I'm surprised there aren't already, though, so maybe I'm missing something. But this makes me uncomfortable - I think it would be better to wait for a proper solution for closure+metadce in this PR.

kripken commented on src/preamble.js:
```diff
@@ -317,10 +317,25 @@ function allocate(slab, types, allocator, ptr) {
+      // Fall through to perform a static allocation if malloc is not yet available
+    default: // case ALLOC_STATIC:
+      ret = staticAlloc(sz);
   }
```
this version is larger, and I don't see the benefit?

juj replied:
The reason for this change is that earlier code referenced `stackAlloc`, which is exported from asm.js/wasm module, and after no longer statically declaring it, the `allocate()` function would fail at undefined reference to `stackAlloc`. This code is larger by ~150 bytes, but this also avoid temp garbage creation, so that is worth it to my view.

I could have added `typeof stackAlloc === 'function'` checks, but that makes the code really complex, and the temp garbage variant there really bugged me.

Ideally in the future we'll work towards removing `allocate()` altogether, and call sites should know what/where they are allocating, so preferably we will do away with this function in the future anyways. I don't think any user code should ever be calling `allocate()`, so this should be doable.

kripken replied:
I see, thanks.

Then this change should no longer be necessary after the fixes for closure/metadce, beecause then we will know stackAlloc exists in those optimizers?

juj replied:
This is still needed - before the code was

```javascript
function allocate(...) {
  ...
  ret = [..., stackAlloc, ...](...);
  ...
}
...
var stackAlloc = Module['stackAlloc'] = asm['stackAlloc'];
```

which parsed, but `stackAlloc` would be `undefined` in all calls to `allocate()` until execution reached to receiving the exports. (If anything attempted to do a `stackAlloc()`, it'd throw an exception trying to call `undefined()`, but that does not happen).

After this change we no longer have the line `var stackAlloc = Module['stackAlloc'] = asm['stackAlloc'];`, at which point referencing `stackAlloc` is not allowed, but throws immediately. I could do `typeof stackAlloc == 'function' ? stackAlloc : 0` there, but thought it is just better to simplify code.

kripken replied:
If we have to, this version with `ifdef`ing seems ok, but could we even more simply do in an earlier toplevel location
```
#if !DECLARE_..
var stackAlloc; // declare it explicitly for allocate()
#endif 
```
That would give it a value (undefined) so allocate does not crash before the asm module returns the implementation?

kripken commented:
The closure issue here does look serious. I think maybe we can fix that in a proper way by *dynamically* adding more externs (added to what we have hardcoded in `src/closure-externs.js` that define Module and those properties on Module. I think we should investigate that before considering adding options here to work around the closure issue.

juj commented:
I did consider the option of dynamically creating an externs file from the module exports, but that looked like more complex data wrangling through emscripten internals that could be a future improvement. The code does disable the size optimization at present altogether when closure is being used, in which case the variables will be generated similar to the old form, statically declaring every export.

The remaining test failures to should not be closure-related externs issues, since the optimization is not currently done with closure, but they look like some other issues.

juj commented:
Updated the code a bit, I think all tests should now hopefully pass.

It looks like the remaining issue is that JS optimizer is hard wired to look for patterns

```js
var _foo1 = Module['_foo1'] = asm['_foo1'];
var _foo2 = Module['_foo2'] = asm['_foo2'];
```

when it is analyzing what got exported from the Wasm module (only in Wasm mode (`binaryen3` suite, i.e. `-O3 -s WASM=1`) it seems, not in asm.js?).

The above format is not ideal, at minimum if we are in a scenario where we do have to statically declare exports, we would like to do it in form

```js
var _foo1 = asm['_foo1'];
var _foo2 = asm['_foo2'];
for(var i in asm) Module[i] = asm[i];
```

because the second form only has two static references to unminified names, versus the first form which has three.

It looks like there are two big follow-up work items to this:
 - build a path from the exports generator code to create a module-closure-externs.js file that tells Closure about all the module exports so that it can ignore them, and
 - can the Wasm JS optimizer be modified to not have to need to see a static list of all exports? If so, then I believe we would never need to list unminified names, but always could do a loop.

kripken commented:
> when it is analyzing what got exported from the Wasm module (only in Wasm mode (binaryen3 suite, i.e. -O3 -s WASM=1) it seems, not in asm.js?).

Makes sense, that is how metadce works. It needs to statically see the imports and exports from the wasm module, so that it can dce it all.

I think this can be solved like with closure: the underlying issue for both is that we need to tell those optimizers the list of names that appear on Module. Once we have that list we just need to provide it to either of those tools in an appropriate form (I can help with adding an interface for metadce).

juj commented:
Updated the PR and added a dynamically generated closure externs .js file so that the optimization can be enabled also when Closure is enabled, so that only leaves the JS optimizer part (-O3 and >= -Os builds). Could the JS optimizer reuse the same externs file as Closure?

This PR should be safe in the sense that the feature is now disabled for -O3 and -Os, and only enabled in smaller opt levels.

Btw, when this is merged, it should not be squashed, since the individual commits are useful standalone and for bisecting through history.

kripken commented on src/jsifier.js:
```diff
@@ -282,6 +282,12 @@ function JSify(data, functionsOnly) {
+          }
+          if (!LibraryManager.library[snippet + '__asm'] && LibraryManager.library[ident + '__asm']) {
+            error('An asm.js/wasm function "' + snippet + '" cannot be constructed as an alias of a non-asm.js/wasm function "' + ident + '"! (the JS function needs to be imported to asm.js/wasm)');
+          }
```
Why is this necessary?

Is it still necessary after we get closure and metadce fixed?

juj replied:
It is wrong to have a `library_foo.js` file with

```javascript
  foo__asm: 'true',
  foo: function() {
    (asm code here)
  },

  foo_alias: 'foo',
```

because that would generate code `var foo_alias = foo;` way before the asm/wasm module has even been loaded, so `foo_alias` ends up being `undefined`, as `foo` is received as an export only later. The above change enforces that one does not alias across non-asmjs and asmjs functions.

kripken commented:
I'm still worried about the complexity here - I asked two questions about specific things above.

Ideally this PR should contain the closure and metadce fixes and the change itself, and none of the other changes (like the chnage to allocate, to library.js, etc.) - that's what I think should be possible, unless I'm missing something?

juj commented:
It looks like the only remaining failing test is

```
======================================================================
FAIL: test_asm_js_function_exports_are_only_referenced_once (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/usr/lib/python3.5/unittest/case.py", line 600, in run
    testMethod()
  File "/root/emscripten/tests/test_other.py", line 8817, in test_asm_js_function_exports_are_only_referenced_once
    self.assertEqual(num_times_export_is_referenced, 1)
  File "/usr/lib/python3.5/unittest/case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.5/unittest/case.py", line 813, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 3 != 1

----------------------------------------------------------------------
Ran 324 tests in 695.434s
```

on `test-upstream-other`, which is something I am unable to reproduce. I have tried on Windows, macOS and Linux, and with python2 and python3, attempted to run in singlethreaded and multithreaded mode, thinking perhaps previous test would confuse here. Is there something special with `test-upstream-other`?

juj commented:
> none of the other changes (like the chnage to allocate, to library.js, etc.)

The change to allocate is needed, and the change to library.js is to add a bug fix that needs to be fixed before the static declarations could be dropped, otherwise code would incorrectly reference undefined variables `memcpy`, `memset` and `memmove` at global scope, which do not yet exist before the asm.js/wasm module exports have been obtained. This was incorrect already before this PR, but the `llvm_mem*` alias functions did not seem to be getting used anywhere, which is why this was never noticed.

juj commented:
> metadce fixes

I tried looking at this, but unfortunately this is not something I am able to do immediately. I was hoping that you might have a moment to help? Or if not, and you still think this cannot land until metadce also works, then I think we should probably put this on hold until I can allocate more time.

kripken commented:
I'll take a look at the metadce stuff here later today.

kripken commented on emscripten.py:
```diff
@@ -1480,10 +1480,27 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
+  # Generate an exports file that records all the exported symbols from asm.js/wasm module.
+  module_exports_suppressions = '\n'.join(['/**\n * @suppress {duplicate, undefinedVars}\n */\nvar %s;\n' % i for i in module_exports])
+  exports_file = get_configuration().get_temp_files().get('_module_exports.js')
+  shared.Settings.MODULE_EXPORTS_FILE = exports_file.name
```
I think we should avoid using a file for this. How about storing the list of wasm/asm exports in a plain list (`MODULE_EXPORTS`, without `_FILE`), and only if we run closure, we generate a file for it in the code about to run closure?

juj replied:
Yeah, that's doable - adjusted.

kripken commented on emscripten.py:
```diff
@@ -1480,10 +1480,27 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
+  exports_file.write(module_exports_suppressions.encode())
+  exports_file.close()
+
   if not shared.Settings.SWAPPABLE_ASM_MODULE:
```
trying to check out what happens with metadce, I noticed that since we actually use `SWAPPABLE_ASM_MODULE` by default for wasm (for async compilation), so this optimization is not occurring for wasm anyhow (except when async compilation is off)

juj replied:
That looks right :/ We can't do this for `SWAPPABLE_ASM_MODULE` it seems. Disabled it for that scenario for now, so when `SWAPPABLE_ASM_MODULE`, then write out the exports in full.

kripken replied:
I missed something - why can't we do it for the swappable case? Might be more complicated to add the indirection, but not impossible, I'd hope, but I didn't try.

if we can't do this for the swappable case, then probably we should not do it for the wasm case, since wasm+sync compilation is not an important mode.

kripken commented on emscripten.py:
```diff
@@ -1480,10 +1480,27 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
+#      receiving += ''.join(['var ' + s + ' = asm["' + s + '"];\n' for s in module_exports])
+#      receiving += 'for(var module_exported_function in asm) Module[module_exported_function] = asm[module_exported_function];\n'
+    else:
+      receiving += '(function() { for(var i in asm) this[i] = Module[i] = asm[i]; }());\n'
```
use of `this` is worrying - you're modifying the global scope here. That would break in MODULARIZE etc., and also is not good for VM optimizations. But I see how it's hard to avoid it to really shrink the size down... hmm

juj replied:
The intent is that in `MODULARIZE` mode, the `this[i]` exports would be assigned to the scope of the `MODULARIZE` module - does that not occur? If so, then I suppose we need to change the assignment to `{{{EXPORT_NAME}}}[i]` or something like that? I don't care for now if this is incompatible with MODULARIZE, that can be improved on later (we do use MODULARIZE, so eventually will need to)

juj replied:
Hmm at least testing locally, I do not find the variable to leak outside the MODULARIZE scope to global scope.

kripken replied:
How did you test? If I build a hello world with MODULARIZE, and then replace the top line which is
```
var Module = (function() {
```
with this:
```
var inGlobalScope = 10;

var Module = (function() {

  console.log(this['inGlobalScope']);
```
then I see that `this` refers to the global scope, outside the modularize function.

We could perhaps work around that, by forcing `this`. But in general it's a bad idea, since it will break in other situations, like if the user applies some function scoping later on.

kripken replied:
I wonder if it may make sense to consider a more moderate approach here, like emitting
```
var name1, name2, .. , nameN;
for (var i in asm) Module[i] = asm[i];
```
That would not reduce all the size, but would still get the majority of the win here, and would avoid the global scope issue.

juj replied:
You're correct - looks like my test was flawed - I compiled hello_world.c with MODULARIZE and poked on that in console, but that did not really get anything.

In `-s MODULARIZE=1` mode, we could do

```
var Module = (function(self) {
...
for(var i in asm) self[i] = Module[i] = asm[i];
...
});
```

and 

```
Module(Module);
```
or
```
Module({});
```

to route the assignments?

kripken commented:
because of the above issue I was not able to fully test metadce, but this is vaguely what I think the code might look like:

```
@@ -2512,20 +2507,34 @@ class Building(object):
     return js_file
 
   # run binaryen's wasm-metadce to dce both js and wasm
   @staticmethod
   def metadce(js_file, wasm_file, minify_whitespace, debug_info):
     logger.debug('running meta-DCE')
     temp_files = configuration.get_temp_files()
     # first, get the JS part of the graph
     txt = Building.js_optimizer_no_asmjs(js_file, ['emitDCEGraph', 'noEmitAst'], return_output=True)
     graph = json.loads(txt)
+    # add exports based on the backend output, that are not present in the JS
+    exports = set()
+    for item in graph:
+      if 'export' in item:
+        exports.add(export['export'])
+    logging.error(str(Settings.MODULE_EXPORTS))
+    for export in Settings.MODULE_EXPORTS:
+      if export not in exports:
+        logging.error('add ' + export)
+        graph.append({
+          'export': export,
+          'name': 'emcc$export$' + export,
+          'reaches': []
+        })
     # ensure that functions expected to be exported to the outside are roots
     for item in graph:
       if 'export' in item:
         export = item['export']
         # wasm backend's exports are prefixed differently inside the wasm
         if Settings.WASM_BACKEND:
           export = '_' + export
         if export in Building.user_requested_exports or Settings.EXPORT_ALL:
           item['root'] = True
     if Settings.WASM_BACKEND:
```

juj commented:
Updated the PR. This now requires https://github.com/WebAssembly/binaryen/pull/1761 to land. That PR enables Wasm -O2 builds to also use DECLARE_ASM_MODULE_EXPORTS=0 option.

Unfortunately Meta-DCE is still incompatible with the option, even after the patch to it earlier. It does not properly generate the DCE graph in emitDCEGraph pass if DECLARE_ASM_MODULE_EXPORTS=0 option is in use. This can be seen in `binaryen3.test_ccall` test, which fails with `stackSave` and `stackRestore` functions getting optimized out. Debug-dumping the DCE graph from emitDCEGraph pass, it prints:

```
}, {
    "root": true,
    "name": "emcc$defun$ccall",
    "reaches": ["emcc$defun$Pointer_stringify", "emcc$defun$getCFunc"]
}, {
```

which should have been

``
}, {
    "root": true,
    "name": "emcc$defun$ccall",
    "reaches": ["emcc$defun$Pointer_stringify", "emcc$defun$getCFunc", "emcc$export$stackRestore", "emcc$export$stackSave"]
}, {
```

After the patch, the DCE graph does properly now have

```
}, {
    "export": "stackRestore",
    "name": "emcc$export$stackRestore",
    "reaches": []
}, {
    "export": "stackSave",
    "name": "emcc$export$stackSave",
    "reaches": []
}, {
```

so it gets the elements there, but looks like the `reaches` array is not generated for other functions that reference to functions from `MODULE_EXPORTS` setting variable, and hence even when `ccall` uses `stackSave()` and `stackRestore()`, those are not in `reaches`, and they do get mis-optimized out.

If meta-DCE could work with this, the optimization should be doable in -O3, -Os and -Oz builds in Wasm, when `SWAPPABLE_ASM_MODULE` is not used. Overall I think a good future direction would be to do async Wasm compilation without needing SWAPPABLE_ASM_MODULE, i.e. only process the exports proper when the wasm module has asynchronously compiled, at which point SWAPPABLE_ASM_MODULE would get deprecated and the fact that this optimization does not work without SWAPPABLE_ASM_MODULE would not matter.

Alternatively we could land this with Wasm side disabled.

juj commented:
> Updated the PR. This now requires [WebAssembly/binaryen#1761](https://github.com/WebAssembly/binaryen/pull/1761) to land. That PR enables Wasm -O2 builds to also use DECLARE_ASM_MODULE_EXPORTS=0 option.

On second thought, don't know what the cadence of updating binaryen is, so adjusted the PR to not depend on this just yet, so this could land separately.

juj commented:
This PR seems to fail on the `test-upstream-` CIs, but I do not understand why. It passes locally for me, and I authored a `WASM_BACKEND` check to the code to disable the DECLARE_ASM_MODULE_EXPORTS==0 optimization there, but it still seems to not like something.

kripken commented:
> In -s MODULARIZE=1 mode, we could do `var Module = (function(self) { for(var i in asm) self[i] = Module[i] = asm[i]; });`

Not quite, and I think that is a serious problem here. You can't access a local scope that way, that is, accessing `x` won't work after setting `self[x]`. What *might* work is to use `with`,
```
function func() {
  with (this) {
    console.log('func');
    this['x'] = 10;
    console.log('access x inside function+with will work:', x);
  }
}
func.call({});
console.log('access x inside function will throw:', x);
```
But this is pretty bad from a JS point of view - `with` is highly unrecommended, and there have been JS engine incompatibilities. It's also very possible JS VMs would not JIT a `with` scope, and we'd need that scope around all our glue JS.

So overall I am very worried about this approach, the main problem being that it requires `with` or such for modularize, and for non-modularize it will break if the user does not leave the code in the global scope. As a result, I think we can at best make this a default-off option with big warning signs around it. However, I'd prefer to not add even that, unless you feel strongly.

But I am optimistic in other directions. The real issue is getting rid of those `var x = Module['x'] = asm['x']` assignments, one per asm.js/wasm export. Some possibilities:

 * If you really really care about code size at all costs, you can use eval, `for (var i in asm) eval(i + ' = Module["' + i + '"] = asm["' + i + '"]')`. `eval` of course also has downsides, but less than global scope and `with`. In particular, at least `eval` would just avoid JITing that specific code itself, but not e.g. the WebGL glue code and everything else.
 * We should do a focused effort to shrink the number of those assignments, as a general effort to move things from JS to asm.js/wasm. Basically knock them down one by one.
 * We can coalesce those exports. Consider what we do for EM_ASMs - we have one export per signature, and we send an integer ID. That adds lookup overhead, but may be worth it for code size.

juj commented:
> You can't access a local scope that way, that is, accessing `x` won't work after setting `self[x]`.

Ohh, right.

> What _might_ work is to use `with`
> But this is pretty bad from a JS point of view

Yeah, that does not look interesting, let's not go there.

> As a result, I think we can at best make this a default-off option with big warning signs around it.

That sounds fine.

> However, I'd prefer to not add even that, unless you feel strongly.

This is the single biggest code size save feature that we cannot ignore, so either we'd land it as an option or I'll end up maintaining it in our fork. In applications that export a large surface area of an API to outside world, (think Ammo.js or Box2D.js style of use), this easily saves around 10KB-100KB of uncompressed JS when there are hundreds of exports. In one project I am building, we have 676 exports, amounting to 89637 bytes of extra space, that I can all just replace with single for loop that takes only 55 bytes.

>   you can use eval

This is unfortunately not an option, CSP is enforced in our case.

>    shrink the number of those assignments

The exports here are not about Emscripten internals being exported, but about user code API functions being exported, which can have really large size. While shrinking some of the internal exports would be nice (every byte counts), that would not move the needle much.

> We can coalesce those exports. Consider what we do for EM_ASMs - we have one export per signature, and we send an integer ID. That adds lookup overhead, but may be worth it for code size.

I think I'd rather keep it disabled in MODULARIZE mode for now, and down the line, if one wants to enable it in MODULARIZE mode, investigate generating accesses to exports via `Module` (or via `asm`). User JS library/pre-JS code will need to be mindful of that if it wants to use the feature with MODULARIZE, but code size will stay smaller without performance downsides.

juj commented:
Updated the code to default to DECLARE_ASM_MODULE_EXPORTS=1, but the `test-upstream-` CIs still fail, and I cannot reproduce it locally. To my understanding there should not be anything left in this PR that could affect adversely. Any chance the failures would reproduce for you locally?

kripken commented:
> This is the single biggest code size save feature that we cannot ignore [..] The exports here are not about Emscripten internals being exported, but about user code API functions being exported

Interesting! Can you elaborate on how those exports are then called - is it from handwritten user JS code? And is that code seen by the emscripten optimizer during compilation, or is it external?

As I said, I'm ok with figuring out a way to land this default-off with big warnings signs. But first I'd like to understand it a little better and make sure we've considered alternatives. Some thoughts on those:

 * The code pattern is `var x = Module['x'] = asm['x'];`. We can turn that into a loop for the `Module['x']` stuff safely (no need for `with` etc., no risks with modularize), leaving `var x = asm['x']`. That might remove about half the code here in a simple way (almost trivially, except for a simple metadce parsing change - I can do that if we go this route).
 * That leaves us with `var x = asm['x'];`, which are very hard to get rid of in a safe way (needs `with` or `eval` or to rely on the special global scope). Alternatively, if we never used `x`, but always used `Module['x']`, we would not need these `var x` at all.
   * Changing to that model would not be hard - we don't need to modify existing code, we'd write a js optimizer pass for it (transforming `x` that accesses the toplevel scope into `Module['x']`). This would be a safe operation.
   * The downside here would be that after closure minification we'd in general be turning `a` (minified `x` etc.) into `b.c` (minified `Module['x']`). So if there are many accesses of those properties, it may not be worth it. We could write a minifier pass that checks that and does the optimal thing for each identifier. But that's why I asked where those accesses are, earlier: note that if all uses are in external user code anyhow, then there is no code size downside.

About those tests, I think the test-upstream breakage is due to unrelated recent stuff. But let's wait til after the US holiday when it can be investigated to be sure.

juj commented:
> Interesting! Can you elaborate on how those exports are then called - is it from handwritten user JS code? And is that code seen by the emscripten optimizer during compilation, or is it external?

The calls are from external code that do not go through emscripten optimizer. This is for the kind of "one engine + N game code applications" kind of pattern, where the game code can be written by hand, against a set of functions exported from Module.

However I cannot say that this would be always done, in another use case for Unity3D, many of the exports from asm.js/wasm are only consumed by JS library, and never by external handwritten code, so would be good to be able to emit the optimal thing for both kinds of symbol usages.

>     * The code pattern is `var x = Module['x'] = asm['x'];`. We can turn that into a loop for the `Module['x']` stuff safely (no need for `with` etc., no risks with modularize), leaving `var x = asm['x']`. That might remove about half the code here in a simple way (almost trivially, except for a simple metadce parsing change - I can do that if we go this route).

Unfortunately that would not be good enough for this use, half of that is still too much code. In one of our use cases we do not modularize, so want to save 100% of the excess size here. (In another use case, we do modularize, but that could be looked at separately. Perhaps there we will not be able to avoid `var x = asm['x']`)

kripken commented:
@juj I see, thanks. Ok, sounds like for your use case you need to use JS tricks like accessing the global scope in unsafe ways. As mentioned before I'd be ok with adding a feature like that, default off, and carefully documented as dangerous. We should also work to make the emscripten.py changes clearly separated.

juj commented:
Check. The option DECLARE_ASM_MODULE_EXPORTS=0 enables the unsafe optimization, and it defaults to DECLARE_ASM_MODULE_EXPORTS=1, so the option is only manually enabled. Added documentation in src/settings.js to mention that the setting DECLARE_ASM_MODULE_EXPORTS=0 can be dangerous to use. The changes to emscripten.py are separated to their own commits, so if this PR is merged without squashing, then the individual things are preserved.

kripken commented on src/library.js:
```diff
@@ -894,10 +894,15 @@ LibraryManager.library = {
+  llvm_memcpy_i64: '_memcpy_js',
+  llvm_memcpy_p0i8_p0i8_i32: '_memcpy_js',
+  llvm_memcpy_p0i8_p0i8_i64: '_memcpy_js',
 
```
IIUC, these are needed to work around the linking problem where memcpy etc. was not linked in, but late in codegen we realize we need it?

If so then this should not be necessary for the LLVM wasm backend, and we can ifdef these lines to make sure they don't exist there (if this were brought in for there, it would be a bug).

But regardless, this seems separate from the topic of the main PR here - how does this relate to not declaring the exports? The linker issue should exist regardless?

juj replied:
This is not because of a linking issue, but because of https://github.com/kripken/emscripten/pull/7505/files#r234345567.

Adjusted the code to omit this indirection when `DECLARE_ASM_MODULE_EXPORTS=0` is not being used, so that we do not pay cost of these aliases when not needed.

kripken commented on emcc.py:
```diff
@@ -1344,6 +1349,17 @@ def check(input_file):
+          shared.Settings.DECLARE_ASM_MODULE_EXPORTS = 1
+          logger.warning('Enabling -s DECLARE_ASM_MODULE_EXPORTS=1 since -s SWAPPABLE_ASM_MODULE=1 is used')
+        # Wasm -O3 builds use Meta-DCE which is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0 option.
+        if options.opt_level >= 3 or options.shrink_level >= 1 and not shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
```
instead of hardcoding assumptions about when metadce is run, how about erroring in the metadce code itself in tools/shared.py?

juj replied:
I had added this check to both places: metadce has

```py
  def metadce(js_file, wasm_file, minify_whitespace, debug_info):
    logger.debug('running meta-DCE')
    assert Settings.DECLARE_ASM_MODULE_EXPORTS, 'Internal error: Meta-DCE is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0, build should have occurred with -s DECLARE_ASM_MODULE_EXPORTS=1 if we reach here!'
```

so this code in emcc.py frontend disables `DECLARE_ASM_MODULE_EXPORTS=0` if it's incompatible, but the metadce code then finally enforces that it doesn't get called with a currently incompatible flag.

kripken commented on emcc.py:
```diff
@@ -1344,6 +1349,17 @@ def check(input_file):
+        # concise form of grabbing exports that does not need to refer to each export individually.
+        if shared.Settings.SWAPPABLE_ASM_MODULE == 1 and not shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
+          shared.Settings.DECLARE_ASM_MODULE_EXPORTS = 1
+          logger.warning('Enabling -s DECLARE_ASM_MODULE_EXPORTS=1 since -s SWAPPABLE_ASM_MODULE=1 is used')
```
the this and the modularize handler could be put close together, behind an `if not shared.Settings.DECLARE_ASM_MODULE_EXPORTS`. I also think it's simpler to error on them instead of warning (warnings might be missed).

juj replied:
If user is specifying conditions where the option `DECLARE_ASM_MODULE_EXPORTS=0` is not possible to be used (e.g. `-O3`), it is not the end of the world to have `DECLARE_ASM_MODULE_EXPORTS` option revert to `=1` state. Code will build, but with an omitted optimization.

In build systems it gets really uncomfortable to have to know about combinations of different flags that are mutually incompatible, because they cannot then be specified orthogonally. (`--separate-asm` and `-s WASM=1` are currently an example of this, would be better for `--separate-asm` to just silently get ignored when `-s WASM=1` is in effect)

Ideally the limitation that MetaDCE is not compatible with `DECLARE_ASM_MODULE_EXPORTS=0` could be lifted in the future (and preamble refactored so that `SWAPPABLE_ASM_MODULE` option would be deprecated as unnecessary? though that's probably another conversation altogether), so users would not need to populate their build systems with special case code of "if -O3 or -Oz and we'd like DECLARE_ASM_MODULE_EXPORTS=0, don't actually specify DECLARE_ASM_MODULE_EXPORTS=0 (TODO: perhaps specify it in the future)". When/if later MetaDCE is improved to support DECLARE_ASM_MODULE_EXPORTS=0, then it will be a backwards breaking change in users' build systems when they combine `DECLARE_ASM_MODULE_EXPORTS=0` with `-O3`, older compilers would start aborting the builds if emcc.py early-aborted here, whereas it would be reasonable for `DECLARE_ASM_MODULE_EXPORTS=0` to just have gotten ignored.

I'd prefer to keep these as warnings if that's ok?

kripken replied:
Ok.

kripken commented on src/settings.js:
```diff
@@ -217,6 +217,16 @@ var SIMD = 0;
+// can have dangerous behavior e.g. if using external minifier tools on the code
+// after the build, later, or if attempting to enclose the generated code in an
+// (anonymous) function/module scope.
+var DECLARE_ASM_MODULE_EXPORTS = 1;
```
Pretty good, how about a little rephrasing though, like this?

> If set to 1, each asm.js/wasm module export is individually declared with a JavaScript "var" definition. This is the simple and recommended approach. However, this does increase code size (especiall if you have many such exports), which can be avoided in an unsafe way by setting this to 0. In that case, no "var" is created for each export, and instead a loop (of constant code size, with no hardcoded list of exports) writes all the exports received into the global scope. Doing so is dangerous since such modifications of the global scope can confuse external JS minifer tools, and also things can break if the scope the code is in is not the global scope (e.g. if you manually enclose them in a function scope).

kripken commented on emcc.py:
```diff
@@ -1326,6 +1331,19 @@ def check(input_file):
+          shared.Settings.DECLARE_ASM_MODULE_EXPORTS = 1
+          logger.warning('Enabling -s DECLARE_ASM_MODULE_EXPORTS=1 since -s SWAPPABLE_ASM_MODULE=1 is used')
+        # Wasm -O3 builds use Meta-DCE which is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0 option.
+        if options.opt_level >= 3 or options.shrink_level >= 1:
```
this hardcodes the conditions for metadce, which may change, in which case we'd have potential inconsistencies. how about adding a method `will_metadce` and use it in the relevant places?

juj replied:
Check, updated the code

kripken commented on src/library.js:
```diff
@@ -993,10 +984,53 @@ LibraryManager.library = {
-  llvm_memset_p0i8_i64: 'memset',
 
+#if DECLARE_ASM_MODULE_EXPORTS
+  llvm_memcpy_i32: '_memcpy',
```
I think these should not have `_`, and it should fail with it, so this seems odd if it passed for you?

juj replied:
I did not run this on my system, but instead just let this run through the CI, where it passes. Indeed I agree the code is wrong here, these should be without `_`. The fact that the CI passes suggests that these are not actually used anywhere? Anyhow, fixed to remove `_` now.

kripken commented on src/library.js:
```diff
@@ -993,10 +984,53 @@ LibraryManager.library = {
+  llvm_memset_i32: '_memset_js',
+  llvm_memset_p0i8_i32: '_memset_js',
+  llvm_memset_p0i8_i64: '_memset_js',
+#endif
```
please add the condition in a comment here at the end, so it's easy to match up the `#if` to `#endif`.

kripken commented on tools/shared.py:
```diff
@@ -2557,10 +2586,12 @@ def metadce(js_file, wasm_file, minify_whitespace, debug_info):
     logger.debug('minifying wasm imports and exports')
     # run the pass
     cmd = [os.path.join(Building.get_binaryen_bin(), 'wasm-opt'), '--minify-imports-and-exports', wasm_file, '-o', wasm_file]
+    # TODO: Change the above to following when binaryen is updated:
```
binaryen has been updated, so I think we can do this now

juj replied:
Oh, perfect!

juj commented:
>If we have to, this version with ifdefing seems ok, but could we even more simply do in an earlier toplevel location

```
#if !DECLARE_..
var stackAlloc; // declare it explicitly for allocate()
#endif 
```

>That would give it a value (undefined) so allocate does not crash before the asm module returns the implementation?

Oh, great idea, that's more compact as well! (Plus `var stackAlloc;` should still Closure out if `allocate()` is closured out)
